adapta/storage/distributed_object_store/datastax_astra/astra_client.py

```diff
@@ -0,0 +1,316 @@
+
+        cloud_config = {"secure_connect_bundle": os.path.join(self._tmp_bundle_path, tmp_bundle_file_name)}
+        auth_provider = PlainTextAuthProvider(self._client_id, self._client_secret)
+
+        profile = ExecutionProfile(
+            retry_policy=RetryPolicy(),
+            consistency_level=ConsistencyLevel.LOCAL_QUORUM,
+            serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
+            request_timeout=self._socket_read_timeout / 1e3,
+            row_factory=dict_factory,
+        )
+
+        # https://docs.datastax.com/en/developer/python-driver/3.28/getting_started/
+        self._session = Cluster(
+            execution_profiles={EXEC_PROFILE_DEFAULT: profile},
+            cloud=cloud_config,
+            auth_provider=auth_provider,
+            reconnection_policy=ExponentialReconnectionPolicy(
+                self._reconnect_base_delay_ms, self._reconnect_base_delay_ms
+            ),
+            compression=True,
+            application_name=self._client_name,
+            application_version=__version__,
+            sockopts=[
+                (IPPROTO_TCP, TCP_NODELAY, 1),
+                (IPPROTO_TCP, TCP_USER_TIMEOUT, self._socket_read_timeout),
+            ],
+        ).connect(self._keyspace)
+
+        set_session(self._session)
+
+        shutil.rmtree(self._tmp_bundle_path)
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._session.shutdown()
+        self._session = None
+
+    def get_table_metadata(self, table_name: str) -> TableMetadata:
+        """
+        Returns Cassandra/Astra table metadata.
+
+        NB. Use the Force, read the Source: https://github.com/datastax/python-driver/blob/master/tests/integration/standard/test_metadata.py#L233-L238
+
+        :param: table_name: Name of the table to read metadata for.
+        """
+        return get_schema_parser(self._session.cluster.control_connection._connection, "4-a", None, 0.1).get_table(
+            keyspaces=None, keyspace=self._keyspace, table=table_name
+        )
+
+    def get_entity(self, table_name: str) -> Dict:
+        """
+        Reads a single row from a table as dictionary
+        https://docs.datastax.com/en/developer/python-driver/3.28/cqlengine/queryset/
+
+        :param: table_name: Name of the table to read a row from.
+        """
+
+        named_table = NamedTable(self._keyspace, table_name)
+        return named_table.objects[0]
+
+    def get_entities_from_query(self, query: str, mapper: Callable[[Dict], TModel]) -> pandas.DataFrame:
+        """
+        Maps query result to a pandas Dataframe using custom mapper
+
+        :param: query: A CQL query to execute.
+        :param: mapper: A mapping function from a Dictionary to the desired model type.
+        """
+        return pandas.DataFrame([mapper(entity) for entity in self._session.execute(query)])
+
+    def filter_entities(
+        self,
+        model_class: TModel,
```

Comment:
Should it be `model_class: Type[TModel]`? It's not clear from the definition should I provide instance of the TModel type or class constructor?

Reply:
good catch, updated